buildpacks/jvm-function-invoker/src/util/logger.rs

```diff
@@ -0,0 +1,86 @@
+use anyhow::anyhow;
+use std::{fmt::Display, io::Write};
+use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+pub struct Logger {
+    debug: bool,
+}
+
+impl Logger {
+    pub fn new(debug: bool) -> Self {
+        Logger { debug }
+    }
+
+    pub fn header(&self, msg: impl Display) -> anyhow::Result<()> {
+        Ok(header(msg)?)
+    }
+
+    pub fn info(&self, msg: impl Display) -> anyhow::Result<()> {
+        Ok(info(msg)?)
+    }
+
+    pub fn error(&self, header: impl Display, msg: impl Display) -> anyhow::Result<()> {
+        Ok(error(header, msg)?)
+    }
+
+    pub fn warning(&self, header: impl Display, msg: impl Display) -> anyhow::Result<()> {
+        Ok(warning(header, msg)?)
+    }
+
+    pub fn debug(&self, msg: impl Display) -> anyhow::Result<()> {
+        Ok(debug(msg, self.debug)?)
+    }
+}
+
+pub fn header(msg: impl Display) -> anyhow::Result<()> {
+    let mut stdout = StandardStream::stdout(ColorChoice::Always);
+    stdout.set_color(ColorSpec::new().set_fg(Some(Color::Magenta)).set_bold(true))?;
+    writeln!(&mut stdout, "\n[{}]", msg)?;
+    stdout.reset()?;
+    stdout.flush()?;
+
+    Ok(())
+}
+
+pub fn info(msg: impl Display) -> anyhow::Result<()> {
+    let mut stdout = StandardStream::stdout(ColorChoice::Always);
+    writeln!(&mut stdout, "[INFO] {}", msg)?;
+    stdout.flush()?;
+
+    Ok(())
+}
+
+pub fn error(header: impl Display, msg: impl Display) -> anyhow::Result<()> {
+    let mut stderr = StandardStream::stderr(ColorChoice::Always);
+    stderr.set_color(ColorSpec::new().set_fg(Some(Color::Red)).set_bold(true))?;
+    writeln!(&mut stderr, "\n[ERROR: {}]", header)?;
+    stderr.set_color(ColorSpec::new().set_fg(Some(Color::Red)))?;
+    writeln!(&mut stderr, "{}", msg)?;
+    stderr.reset()?;
+    stderr.flush()?;
+
+    Err(anyhow!(format!("{}", header)))
+}
+
+pub fn debug(msg: impl Display, debug: bool) -> anyhow::Result<()> {
+    if debug {
+        let mut stdout = StandardStream::stdout(ColorChoice::Always);
+        writeln!(&mut stdout, "[DEBUG] {}", msg)?;
+        stdout.flush()?;
+    }
+
+    Ok(())
+}
+
+pub fn warning(header: impl Display, msg: impl Display) -> anyhow::Result<()> {
+    let mut stdout = StandardStream::stdout(ColorChoice::Always);
+    stdout.set_color(ColorSpec::new().set_fg(Some(Color::Yellow)).set_bold(true))?;
+    writeln!(&mut stdout, "\n[WARNING: {}]", header)?;
+    stdout.flush()?;
+    stdout.set_color(ColorSpec::new().set_fg(Some(Color::Yellow)))?;
+    writeln!(&mut stdout, "{}", msg)?;
+    stdout.reset()?;
+    stdout.flush()?;
+
+    Ok(())
+}
```

Comment:
Worth mentioning that I like this pattern of having warnings take and format a link/url to a devcenter doc explicitly. I think it's a good practice to always have docs for every warning and error. Having the tools support this natively encourages this practice https://github.com/heroku/buildpacks-ruby/blob/9b8c6c43ca3bd53bb8f915aa0a17636898c1d2c6/lib/heroku_buildpack_ruby/user_comms.rb#L117-L124.

Not a :todo: but it's worth talking about in future iterations.

Reply:
> I think it's a good practice to always have docs for every warning and error

So I'm on the fence about this. Whilst some errors should definitely have docs, for me:
(a) the need for docs about the error shows that the product itself is confusing or the error message doesn't explain enough
(b) docs get out of date and are harder to keep in sync with deployments (particularly when if we start doing things like staggered releases)

That is to say I think most error messages should attempt to contain enough detail that docs aren't needed. And if they do need to link to something, it should be instructions about how to use the product (eg linking to a page that explains the Python buildpack's Django support), rather than a section of the docs that talks about specific error messages.

Reply:
I imagine the languages team will make a much better heroku/salesforce specific logging library than the one I've done.